pkg/validation/internal/operatorhub.go

```diff
@@ -81,41 +101,202 @@ func validateBundleOperatorHub(bundle *manifests.Bundle) errors.ManifestResult {
+		minKube := bundle.CSV.Spec.MinKubeVersion
+		if len(strings.Split(bundle.CSV.Spec.MinKubeVersion, ".")) == 2 {
+			minKube += ".0"
 		}
-		if maintainer.Email != "" {
-			_, err := mail.ParseAddress(maintainer.Email)
-			if err != nil {
-				errs = append(errs, fmt.Errorf("csv.Spec.Maintainers email %s is invalid: %v", maintainer.Email, err))
+		var err error
+		if semverMinKube, err = semver.Make(minKube); err != nil {
+			errs = append(errs, fmt.Errorf("unable to use csv.Spec.MinKubeVersion to verify the CRD/Webhook apis "+
+				"because it has an invalid value: %s", bundle.CSV.Spec.MinKubeVersion))
+		}
+	}
+
+	// If the k8sVersion was informed then, the check is against this version (it means that it is to verify
+	// the operator bundle against the k8s version which it is intended to be distributed).
+	// However, if the minKubeVersion not be informed then, it means that the operator can be distributed in any version.
+	// To ensure this scenario, when the k8sVersion is not informed, the check is made and a warning is raised
+	// if the deprecated/unsupported APIs are found. See that the error only can be safely raised when the
+	// minKubeVersion value is >=1.22.
+	if !hasK8sVersion || semverK8sVersion.GE(k8s1_16) {
+		isUnsupported := semverK8sVersion.GE(k8s1_22) || semverMinKube.GE(k8s1_22)
+
+		// Note that if the operator bundle has both versions then, it will only raise a warning since it might
+		// can means that the operator has been covering both scenarios.
+		if hasDeprecatedAPIs(bundle) {
+			if isUnsupported && !hasV1(bundle) {
+				errs = append(errs, fmt.Errorf(apiDeprecationMsg))
+			} else {
+				warns = append(warns, fmt.Errorf(apiDeprecationMsg))
 			}
 		}
 	}
 
-	for _, link := range csv.Spec.Links {
-		if link.Name == "" || link.URL == "" {
-			errs = append(errs, fmt.Errorf("csv.Spec.Links elements should contain both name and url"))
+	return errs, warns
+}
+
+// hasDeprecatedAPIs returns true when the bundle has one CRD/Webhook using v1beta1
+func hasDeprecatedAPIs(bundle *manifests.Bundle) bool {
+	if len(bundle.V1beta1CRDs) > 0 {
+		return true
+	}
+	for _, obj := range bundle.CSV.Spec.WebhookDefinitions {
+		for _, v := range obj.AdmissionReviewVersions {
+			if v == "v1beta1" {
+				return true
+			}
 		}
-		if link.URL != "" {
-			_, err := url.ParseRequestURI(link.URL)
+	}
+	return false
+}
+
+// hasV1 returns true when the bundle has a CRD/Webhook using V1
+func hasV1(bundle *manifests.Bundle) bool {
+	if len(bundle.V1CRDs) > 0 {
+		return true
+	}
+	for _, obj := range bundle.CSV.Spec.WebhookDefinitions {
+		for _, v := range obj.AdmissionReviewVersions {
+			if v == "v1" {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// validateHubCSVSpec will check the CSV against the criteria to publish an
+// operator bundle in the OperatorHub.io
+func validateHubCSVSpec(csv v1alpha1.ClusterServiceVersion) ([]error, []error) {
+	var errs []error
+	var warns []error
+
+	errs = checkSpecProviderName(csv, errs)
```

Comment:
I feel these functions are very similar so maybe we should make their return signature identical (return errs and warns). Sure some of them don't have warnings for now but maybe they will in the future? This is just a nit. I don't want to drag this down as it is not important.

Reply:
done